File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        int open = n;
        int close = n;
        String cur = "";
        List<String> result = new ArrayList<>();

        dfs(open, close, cur, result);

        return result;
    }

    private void dfs(int open, int close, String cur, List<String> result) {
        if (open == 0 && close == 0) {
            result.add(cur);
        } else {
            if (open > 0) {
                dfs(open - 1, close, cur + "(", result);
            }

            if (close > open) {
                dfs(open, close - 1, cur + ")", result);
            }
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.kt
class Solution {
    fun generateParenthesis(n: Int): List<String> {
        val result = mutableListOf<String>()

        fun dfs(open: Int, close: Int, cur: String) {
            if (open == 0 && close == 0) {
                result.add(cur)
                return
            }
            if (open > 0) dfs(open - 1, close, "$cur(")
            if (close > open) dfs(open, close - 1, "$cur)")
        }

        dfs(n, n, "")

        return result
    }
}